3. Import the required packages with their customary aliases as follows:

    `import pandas as pd`   
    `import numpy as np`  
    `import matplotlib.pyplot as plt`  
    `import seaborn as sns`

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

4.	Using the pandas `read_csv()` method, read the GDP dataset into your notebook as a DataFrame called `gdp_df`. 
    Take a look at the first few and last few rows to familiarize yourself with what is contained in this dataset.

In [2]:
gdp_df = pd.read_csv('..\data\da16-python-gdp-and-internet-use-analysis-DSVaughan266\data\gdp_per_capita.csv')

In [3]:
gdp_df.head()

Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2023  1992.424394              NaN
1     Afghanistan  2022  1981.710168              NaN
2     Afghanistan  2021  2144.166570              NaN
3     Afghanistan  2020  2769.685745              NaN
4     Afghanistan  2019  2927.245144              NaN

In [4]:
gdp_df.tail()

Country or Area  Year        Value  Value Footnotes
8214        Zimbabwe  1994  3965.730986              NaN
8215        Zimbabwe  1993  3634.750494              NaN
8216        Zimbabwe  1992  3649.891947              NaN
8217        Zimbabwe  1991  4126.405247              NaN
8218        Zimbabwe  1990  4013.299059              NaN

5.  How many rows and columns are in `gdp_df`? Rows: 8219 Columns: 4 
    What are the data types of each column? object, int64, float64, float64

In [5]:
gdp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8219 entries, 0 to 8218
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  8219 non-null   object 
 1   Year             8219 non-null   int64  
 2   Value            8219 non-null   float64
 3   Value Footnotes  0 non-null      float64
dtypes: float64(2), int64(1), object(1)
memory usage: 257.0+ KB


In [6]:
gdp_df.describe()

Year          Value  Value Footnotes
count  8219.000000    8219.000000              0.0
mean   2006.672345   20270.976242              NaN
std       9.782986   22532.004246              NaN
min    1990.000000     510.822823              NaN
25%    1998.000000    4284.868430              NaN
50%    2007.000000   11916.457927              NaN
75%    2015.000000   28543.305542              NaN
max    2023.000000  174339.077325              NaN

6.  Drop the `Value Footnotes` column.

In [7]:
gdp_df_droppedVF = gdp_df.drop("Value Footnotes", axis = 1)

In [8]:
print(gdp_df_droppedVF.columns)

Index(['Country or Area', 'Year', 'Value'], dtype='object')


6.  Rename the remaining three to 'Country', 'Year', and 'GDP_Per_Capita'.

In [9]:
gdp_df_droppedVF.columns = ["Country", "Year", "GDP_Per_Capita"]

In [10]:
gdp_df_droppedVF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8219 entries, 0 to 8218
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         8219 non-null   object 
 1   Year            8219 non-null   int64  
 2   GDP_Per_Capita  8219 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 192.8+ KB


In [11]:
gdp_df_clean = gdp_df_droppedVF

7.  How many countries have data for all years? Which countries are missing many years of data? Look at the number of observations per year. 
    What do you notice? 

In [12]:
gdp_df_clean.groupby("Country").size()
# There are 247 countries

Country
Afghanistan                    24
Africa Eastern and Southern    34
Africa Western and Central     34
Albania                        34
Algeria                        34
                               ..
Virgin Islands                 21
West Bank and Gaza             30
World                          34
Zambia                         34
Zimbabwe                       34
Length: 247, dtype: int64

In [13]:
# Investigate the range of the data.
first_year = gdp_df_clean["Year"].min()
last_year = gdp_df_clean["Year"].max()
total_years = last_year - first_year + 1   

In [14]:
print("First year:", first_year)
print("Last_year:", last_year)
print("Total_years:", total_years)

First year: 1990
Last_year: 2023
Total_years: 34


In [15]:
# How many reports were made for each year?
year_counts = gdp_df_clean["Year"].value_counts()
print(year_counts)

Year
2020    247
2021    247
2013    247
2016    247
2019    247
2018    247
2017    247
2014    247
2015    247
2022    246
2011    246
2012    246
2009    245
2010    245
2008    244
2007    242
2006    242
2005    241
2003    241
2004    241
2002    241
2023    240
2001    240
2000    240
1999    238
1998    238
1997    238
1996    236
1995    236
1994    235
1993    234
1992    234
1991    234
1990    233
Name: count, dtype: int64


In [17]:
# There are 34 years of reported data. Some countries reported every year. How many and which ones?
country_counts =gdp_df_clean.groupby("Country").size()

In [18]:
countries_with_34 = country_counts[country_counts == 34].index.tolist()
print(countries_with_34)
len(countries_with_34)

['Africa Eastern and Southern', 'Africa Western and Central', 'Albania', 'Algeria', 'Andorra', 'Angola', 'Antigua and Barbuda', 'Arab World', 'Argentina', 'Armenia', 'Aruba', 'Australia', 'Austria', 'Azerbaijan', 'Bahrain', 'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin', 'Bermuda', 'Bolivia', 'Bosnia and Herzegovina', 'Botswana', 'Brazil', 'Brunei', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cabo Verde', 'Cambodia', 'Cameroon', 'Canada', 'Caribbean small states', 'Central African Republic', 'Central Europe and the Baltics', 'Chad', 'Chile', 'China', 'Colombia', 'Comoros', 'Congo', 'Costa Rica', 'Croatia', 'Cyprus', 'Czechia', "Côte d'Ivoire", 'Dem. Rep. Congo', 'Denmark', 'Dominica', 'Dominican Republic', 'Early-demographic dividend', 'East Asia & Pacific', 'East Asia & Pacific (IDA & IBRD)', 'East Asia & Pacific (excluding high income)', 'Ecuador', 'Egypt', 'El Salvador', 'Equatorial Guinea', 'Estonia', 'Eswatini', 'Ethiopia', 'Euro area', 'Europe & Central Asia', 'Eu

228

There are 247 countries in the database with 228 of them reporting all 34 years.

7 (continued) - Which countries are missing many years of data? Look at the number of observations per year. What do you notice?

In [21]:
countries_under_34 = country_counts[country_counts < 34].index.tolist()
print(countries_under_34)
len(country_counts[country_counts < 34])

['Afghanistan', 'Bhutan', 'Cayman Islands', 'Curaçao', 'Djibouti', 'Faroe Islands', 'Greenland', 'Kosovo', 'Lebanon', 'Montenegro', 'San Marino', 'Serbia', 'Sint Maarten (Dutch part)', 'Somalia', 'Syrian Arab Republic', 'Tonga', 'Turks and Caicos Islands', 'Virgin Islands', 'West Bank and Gaza']


19

In [25]:
# Save countries_under_34 as a dataframe named countries_not_reporting.
countries_not_reporting = gdp_df_clean[gdp_df_clean["Country"].isin(countries_under_34)]

In [29]:
# Create a pivot table to explore the 19 countries that did not report every year.
pd.set_option("display.max_columns", None)
countries_not_reporting.pivot_table(values = "GDP_Per_Capita", index = "Country", columns = "Year")

Year                               1990          1991          1992  \
Country                                                               
Afghanistan                         NaN           NaN           NaN   
Bhutan                      3411.383758   3338.009421   3624.597815   
Cayman Islands                      NaN           NaN           NaN   
Curaçao                            NaN           NaN           NaN   
Djibouti                            NaN           NaN           NaN   
Faroe Islands                       NaN           NaN           NaN   
Greenland                  37619.556421  37644.577619  35849.487074   
Kosovo                              NaN           NaN           NaN   
Lebanon                     5345.512097   7828.172011   8924.152646   
Montenegro                          NaN           NaN           NaN   
San Marino                          NaN           NaN           NaN   
Serbia                              NaN           NaN           NaN   
Sint Maarten (Dutch part)           NaN           NaN           NaN   
Somalia                             NaN    836.966662    803.498728   
Syrian Arab Republic        5828.405516   6081.089071   6662.243142   
Tonga                       4487.819252   4766.388081   4772.330988   
Turks and Caicos Islands            NaN           NaN           NaN   
Virgin Islands                      NaN           NaN           NaN   
West Bank and Gaza                  NaN           NaN           NaN   

Year                               1993          1994          1995  \
Country                                                               
Afghanistan                         NaN           NaN           NaN   
Bhutan                      3850.551289   4042.494178   4379.870466   
Cayman Islands                      NaN           NaN           NaN   
Curaçao                            NaN           NaN           NaN   
Djibouti                            NaN           NaN           NaN   
Faroe Islands                       NaN           NaN           NaN   
Greenland                  34108.678362  35934.108534  37073.577832   
Kosovo                              NaN           NaN           NaN   
Lebanon                     9695.315492  10294.812771  10762.442696   
Montenegro                          NaN           NaN           NaN   
San Marino                          NaN           NaN           NaN   
Serbia                              NaN           NaN   9436.235802   
Sint Maarten (Dutch part)           NaN           NaN           NaN   
Somalia                      812.265795    634.500032    636.433676   
Syrian Arab Republic        6861.795047   7142.953801   7431.354210   
Tonga                       4946.717403   5188.439284   5508.339144   
Turks and Caicos Islands            NaN           NaN           NaN   
Virgin Islands                      NaN           NaN           NaN   
West Bank and Gaza                  NaN   3966.688751   4062.962903   

Year                               1996          1997          1998  \
Country                                                               
Afghanistan                         NaN           NaN           NaN   
Bhutan                      4616.919791   4778.705118   4965.876461   
Cayman Islands                      NaN           NaN           NaN   
Curaçao                            NaN           NaN           NaN   
Djibouti                            NaN           NaN           NaN   
Faroe Islands                       NaN           NaN           NaN   
Greenland                  37573.333407  38057.177195  40936.545081   
Kosovo                              NaN           NaN           NaN   
Lebanon                    11757.282489  11668.400685  11885.307228   
Montenegro                          NaN  14605.661371  15368.186479   
San Marino                          NaN  66923.030342  71024.559499   
Serbia                     10025.965885  10955.595278  11587.873271   
Sint Maarten (Dutch pa

These countries didn't begin reporting until atleast 2000, some as late as 2013:
Afghanistan, Cayman Islands, CuraAsao, Djibouti, Faroe Islands, Sint Maarten, Turks & Caicos, Virgin Islands

In [ ]:
# 8.  In this question, you're going to create some plots to show the distribution of GDP per capita for the year 2020. 
#     Go to the Python Graph Gallery (https://www.python-graph-gallery.com/) and look at the different types of plots under the Distribution section. 
#     Create a histogram, a density plot, a boxplot, and a violin plot. What do you notice when you look at these plots? 
#     How do the plots compare and what information can you get out of one type that you can't necessarily get out of the others?

# 9.  What was the median GDP per capita value in 2020?

# 10. For this question, you're going to create some visualizations to compare GDP per capita values for the years 1990, 2000, 2010, and 2020. 
#     Start by subsetting your data to just these 4 years into a new DataFrame named gdp_decades. Using this, create the following 4 plots:
# 	* A boxplot
# 	* A barplot (check out the Barplot with Seaborn section: https://www.python-graph-gallery.com/barplot/#Seaborn)
# 	* A scatterplot
# 	* A scatterplot with a trend line overlaid (see this regplot example: https://www.python-graph-gallery.com/42-custom-linear-regression-fit-seaborn)  

# Comment on what you observe has happened to GDP values over time and the relative strengths and weaknesses of each type of plot.
# 11. Which country was the first to have a GDP per capita greater than $100,000?
# 
# 12. Which country had the highest GDP per capita in 2020? Create a plot showing how this country's GDP per capita has changed over the timespan 
#     of the dataset.
 
# 13. Which country had the lowest GDP per capita in 2020? Create a plot showing how this country's GDP per capita has changed over the timespan 
#     of the dataset. **Bonus question:** Is it true in general that coutries had a higher GDP per capita in 2020 than in 1990? 
#     Which countries had lower GDP per capita in 2020 than in 1990?

# 14. Read in the internet use dataset into a DataFrame named `internet_df`. You will likely get errors when doing this. 
#     Check the arguments for the read_csv function to find ones that can help correct the errors 
#     (https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) Once you are able to read it in, take per a look at the top and bottom 
#     few rows to make sure that it has been read in correctly. Also, check the datatypes of the columns.

# 15. Drop the `Value Footnotes` column and rename the remaining three to 'Country', 'Year', and 'Internet_Users_Pct'.

# 16. Look at the number of observations in this dataset per year. What do you notice?

# 17. What is the first year to have a non-zero internet users percentage value?

# 18. How does the distribution of internet users percent differ for 2000 and 2014?

# 19. For how many countries was the percentage of internet users below 5% in 2014?

# 20. Merge the two DataFrames to one. Do this in a way that keeps **all rows** from each of the two DataFrames. 
#     Call the new DataFrame `gdp_and_internet_use`. Look at the first and last few rows to confirm that it merged correctly.

# 21. Find the three countries with the highest internet users percentage in 2014. Use a seaborn FacetGrid 
#     (https://seaborn.pydata.org/generated/seaborn.FacetGrid.html) to compare how the GDP per capita has changed over time for these three countries. 
#     What do you notice?

# 22. Subset `gdp_and_internet_use` to just the year 2014. Save this as a new dataframe named `gdp_and_internet_use_2014`.

# 23. Create a plot which compares Internet Users Percentage and GDP per Capita for the year 2014. What do you notice from this plot? 
#     If you see any unusual points, investigate them.

# 24. **Stretch Question:** Use the `qcut` function from pandas (https://pandas.pydata.org/docs/reference/api/pandas.qcut.html) to divide countries 
#     in `gdp_per_capita_2014` into three groups based on their GDP per capita values. Label these groups as "Low", "Medium", and "High". 
#     Put these labels in a new column, named "GDP_group".

# 25. **Stretch Question:** How does the median internet users percentage compare for the three gdp groups?

# ### Bonus exercise:
# 1.    Download another data set from the UN data (http://data.un.org/Explorer.aspx) to merge with your data and explore.
